### PR TITLE
Add GAT related tests

### LINKED ISSUE #4969: Add any relevant test cases for reported GAT issues
We've had a few, and occasionally duplicative, reports of GAT related issues. To the best of my knowledge all the ones we're currently aware of have been resolved but I can't recall if we've added test cases for everything that's been reported. More generally, we could probably use some _more_ tests beyond what's been reported anyway.

For anyone interested in working on this, the tasks will largely consist of:

* reviewing the various GAT-related reports and checking the current tests (advise looking under `tests/target` as not all test inputs have a pair file under `tests/source`). If we're missing cases for any issues and/or if the tests need to be augmented then add/update the test files as needed
* consider whether there's any additional, untested use cases that should be added. Don't feel obligated to come up with something, but would be good take a look because if there's any remaining gaps/bugs it'd be better to catch them proactively vs. waiting for users to run into them. Remember that context matters when formatting, so e.g. it's possible for rustfmt to handle something correctly in a trait context but have a bug in an impl context.

See https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#create-test-cases for more info on adding test cases

### PR DESCRIPTION
**Pull request by cowboy-bebug**
Hi there 👋  I wanted to tackle this task as a first time contributor.

Adding GAT-related tests for the following issues:
- https://github.com/rust-lang/rustfmt/issues/4257
- https://github.com/rust-lang/rustfmt/issues/4322
- https://github.com/rust-lang/rustfmt/issues/4943
- https://github.com/rust-lang/rustfmt/issues/4954

Thanks!
Closes #4969.

### CODE REVIEW

File: tests/target/issue-4257.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    type Type<'a>
+    where
+        T: 'a;
+    fn foo(x: &T) -> Self::Type<'_>;
+}
+impl<T> Trait<T> for () {
+    type Type<'a>
+    where
+        T: 'a,
+    = &'a T;
```
Comment (cowboy-bebug) [inline tests/target/issue-4257.rs:14]: This is the correct behaviour right? Here's another similar example:
https://github.com/rust-lang/rustfmt/blob/e81c393663daab11238944cddfbd97541630f178/tests/target/issue-4159.rs#L15-L17

**Comment by calebcartwright** [comment]
Thanks! Could you make sure none of those files already exist with `_` in the filename instead of `-` (e.g. `issue_4943.rs` vs `issue-4943.rs`)? I feel like there's probably some overlap.

> This is the correct behaviour right? Here's another similar example:

Yes indeed, the [Rust Style Guide](https://github.com/rust-dev-tools/fmt-rfcs) prescribes this formatting for where clauses very explicitly [here](https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/items.md#where-clauses).

**Comment by cowboy-bebug** [comment]
@calebcartwright Yep there aren't any duplicates. Do you want me to add a commit to `mv` all `issue_XXXX.rs` -> `issue-XXXX.rs`?

**Comment by calebcartwright** [comment]
> Do you want me to add a commit to mv all issue_XXXX.rs -> issue-XXXX.rs?

Thank you for the offer but no let's not. That's because I actually want our test files to follow the same `_` file name convention that we follow for our own source files and which in my experience has been the convention in the other rust-lang/* projects. A change like that would also be really noisy and would be better suited to its own PR.

> @calebcartwright Yep there aren't any duplicates.

Oh I see, you've renamed the file for 4943. Please keep the `_` as is with existing files, and update the names of the new files added here with `_` as well. Apologies, I know we don't have this documented and currently have a mix with a greater portion of `-` but since we're already on the topic let's stick with the strategic target of `_` :smile:

**Comment by cowboy-bebug** [comment]
The files are all snake_cased now 😄

**Review by calebcartwright — APPROVED** [review]
Awesome, thank you!

**Pull request merged**